sympy/stats/frv.py

```diff
@@ -375,12 +372,120 @@ class SingleFinitePSpace(SinglePSpace, FinitePSpace):
     def domain(self):
         return SingleFiniteDomain(self.symbol, self.distribution.set)
 
+    @property
+    def _is_symbolic(self):
+        """
+        Helper property to check if the distribution
+        of the random variable is symbolic.
+        """
+        return isinstance(self.domain.set, Intersection)
+
+    def _is_logical(self, expr):
+        """
+        Helper function to check if an expression
+        is logical.
+
+        Returns
+        =======
+
+        True
+            If the expression is logical.
+        False
+            In all other cases.
+        """
+        return isinstance(expr, (Relational, Logic))
+
+    @property
+    def distribution(self):
+        return self.args[1]
+
+    def pdf(self, expr):
+        return self.distribution.pdf(expr)
+
     @property
     @cacheit
     def _density(self):
         return dict((FiniteSet((self.symbol, val)), prob)
                     for val, prob in self.distribution.dict.items())
 
+    @cacheit
+    def compute_characteristic_function(self, expr):
+        if self._is_symbolic:
+            d = self.compute_density(expr)
+            t = Dummy('t', real=True)
+            ki = Dummy('ki')
+            return Lambda(t, Sum(d(ki)*exp(I*ki*t), (ki, self.args[1].low, self.args[1].high)))
+        expr = rv_subs(expr, self.values)
+        return FinitePSpace(self.domain, self.distribution).compute_characteristic_function(expr)
+
+    @cacheit
+    def compute_moment_generating_function(self, expr):
+        if self._is_symbolic:
+            d = self.compute_density(expr)
+            t = Dummy('t', real=True)
+            ki = Dummy('ki')
+            return Lambda(t, Sum(d(ki)*exp(ki*t), (ki, self.args[1].low, self.args[1].high)))
```

Comment:
Is this expression not also valid for the non-symbolic case?

Reply:
There are certain reasons(mentioned below), to separate the computations of symbolic and non-symbolic case,

1. The current API supports non-symbolic dimensions and if we suddenly shift from `dict` to something like, `Lambda`, it will break the tests and hence, the current code of many packages dependent on `sympy.stats.frv`.

2. The current implementation of `Range` doesn't support the use of symbolic parameters which forces me to use extra parameters like, `low` and `high` for symbolic dimensions. As you know, I started some work for allowing symbolic dimensions to `Range` which was terminated due to absence of unsatisfactory reasons to continue the work.

So, I think we can go ahead with this kind of design(I believe it is better than the previous one, though some brute forcing is done here and there which I was compelled to do) for now until we find a very nice solution. May be I can raise on issue on this topic for discussing a nice design which doesn't break the current API as well as allow symbolic dimensions.

Reply:
>     2\. terminated due to absence of unsatisfactory reasons to continue the work.

I don't understand what that means. Clearly there is a reason to have `Range(n)` for a symbol n as demonstrated here.

Reply:
Yes, it can be added to one of the use cases of symbolic `Range`.  I am commenting this at https://github.com/sympy/sympy/issues/16833 for view of other contributors who were participating in the discussion.

Reply:
Another work around can be is to support `sup` and `inf` in `Intersection` which can simplify the things and may be we can avoid making `Range` symbolic.

Reply:
I don't understand why we'd want to avoid making Range symbolic.

Reply:
May be it causes problems like getting the symbolic length, at least I know of this problem, then we decided to drop the work by closing the PR.

Reply:
Let's keep it like this for now. Range can be extended to support symbolic ranges in a future step.